Cython/Utility/CConvert.pyx

```diff
@@ -130,3 +130,74 @@ cdef inline tuple {{to_tuple_cname}}({{base_type}} *v, Py_ssize_t length):
         Py_INCREF(value)
         PyTuple_SET_ITEM(t, i, value)
     return t
+
+#################### c_array_slice.to_py ####################
+
+cdef extern from *:
+    void Py_INCREF(object o)
+    void Py_DECREF(object o)
+    object PyErr_Format(exc, const char *format, ...)
+    int __Pyx_is_valid_index(Py_ssize_t i, Py_ssize_t limit)
+
+    list PyList_New(Py_ssize_t size)
+    void PyList_SET_ITEM(object p, Py_ssize_t pos, object o)
+
+    tuple PyTuple_New(Py_ssize_t size)
+    void PyTuple_SET_ITEM(object p, Py_ssize_t pos, object o)
+
+    set PySet_New(void *)
+    int PySet_Add(object p, object o) except -1
+
+
+@cname("{{cname}}")
+cdef inline {{to_py_obj_type}} {{cname}}({{base_type}} *arr, Py_ssize_t arr_size, Py_ssize_t from_idx, Py_ssize_t to_idx, int wraparound, int boundscheck):
+    cdef Py_ssize_t i, length = 0
+
+    if wraparound:
+        if from_idx < 0: 
+            from_idx += arr_size
+
+        if to_idx < 0: 
+            to_idx += arr_size
+
+    if boundscheck:
+        if from_idx > to_idx:
+             from_idx = to_idx
+        else:
+            if from_idx < 0:
+                from_idx = 0
+
+            if to_idx > arr_size:
+                to_idx = arr_size
+
+    length = to_idx - from_idx
+
+    {{if to_py_obj_type == 'list'}}
+    l = PyList_New(length)
+    {{elif to_py_obj_type == 'tuple' or to_py_obj_type == 'frozenset'}}
+    l = PyTuple_New(length)
+    {{elif to_py_obj_type == 'set'}}
+    l = PySet_New(NULL)
+    {{endif}}
+
+    cdef object value
+    if length > 0:
+        for i in range(<size_t> from_idx, <size_t> to_idx):
```

Comment:
You are mixing signed (`i`) and unsigned types here. That's rarely a good idea in C. Just use `Py_ssize_t` for everything.